atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java

```diff
@@ -0,0 +1,129 @@
+ * MetricsInvocationEventHandler.
+ */
+public class TaggedMetricsInvocationEventHandler extends AbstractInvocationEventHandler<InvocationContext> {
+
+    private static final Logger logger = LoggerFactory.getLogger(TaggedMetricsInvocationEventHandler.class);
+
+    private final TaggedMetricRegistry taggedMetricRegistry;
+    private final String serviceName;
+
+    private final Function<InvocationContext, Map<String, String>> tagFunction;
+
+    public TaggedMetricsInvocationEventHandler(
+            TaggedMetricRegistry taggedMetricRegistry,
+            String serviceName) {
+        this(taggedMetricRegistry, serviceName, unused -> ImmutableMap.of());
+    }
+
+    public TaggedMetricsInvocationEventHandler(
+            TaggedMetricRegistry taggedMetricRegistry,
+            String serviceName,
+            Function<InvocationContext, Map<String, String>> tagFunction) {
+        super(getEnabledSupplier(serviceName));
+        this.taggedMetricRegistry = checkNotNull(taggedMetricRegistry, "metricRegistry");
+        this.serviceName = checkNotNull(serviceName, "serviceName");
+        this.tagFunction = tagFunction;
+    }
+
+    private static String failuresMetricName() {
+        return "failures";
+    }
+
+    private static BooleanSupplier getEnabledSupplier(final String serviceName) {
+        return InstrumentationProperties.getSystemPropertySupplier(serviceName);
+    }
+
+    @Override
+    public InvocationContext preInvocation(Object instance, Method method, Object[] args) {
+        return DefaultInvocationContext.of(instance, method, args);
+    }
+
+    @Override
+    public void onSuccess(@Nullable InvocationContext context, @Nullable Object result) {
+        if (context == null) {
+            logger.debug("Encountered null metric context likely due to exception in preInvocation");
+            return;
+        }
+
+        long nanos = System.nanoTime() - context.getStartTimeNanos();
+        MetricName finalMetricName = MetricName.builder()
+                .safeName(MetricRegistry.name(serviceName, context.getMethod().getName()))
+                .putAllSafeTags(tagFunction.apply(context))
+                .build();
+        taggedMetricRegistry.timer(finalMetricName)
+                .update(nanos, TimeUnit.NANOSECONDS);
+    }
+
+    @Override
+    public void onFailure(@Nullable InvocationContext context, @Nonnull Throwable cause) {
+        markGlobalFailure();
+        if (context == null) {
+            logger.debug("Encountered null metric context likely due to exception in preInvocation: {}",
+                    UnsafeArg.of("exception", cause),
+                    cause);
+            return;
+        }
+
+        String failuresMetricName = MetricRegistry.name(getBaseMetricName(context), failuresMetricName());
+        taggedMetricRegistry.meter(MetricName.builder().safeName(failuresMetricName).build()).mark();
```

Comment:
Should the failure metrics also have the tags applied?  (here, line below and in `markGlobalFailure`)

Reply:
Yep, good spot - that does make sense.